qiskit/transpiler/preset_passmanagers/plugin.py

```diff
@@ -0,0 +1,256 @@
+     - The output of this stage is expected to have every operation be a native
+        instruction on the target backend.
+   * - ``optimization``
+     - ``qiskit.transpiler.optimization``
+     - There are no reserve plugin names
+     - This stage is expected to perform and optimization and simplification.
+       The constraints from earlier stages still apply to the output of this
+       stage. After the ``optimization`` stage is run we expect the circuit
+       to still be executable on the target.
+   * - ``scheduling``
+     - ``qiskit.transpiler.scheduling``
+     - ``alap``, ``asap``
+     - This is the last stage run and it is expected to output a scheduled
+       circuit such that all idle periods in the circuit are marked by explicit
+       :class:`~qiskit.circuit.Delay` instructions.
+
+Writing Plugins
+===============
+
+To write a pass manager stage plugin there are 2 main steps. The first step is
+to create a subclass of the abstract plugin class
+:class:`~.PassManagerStagePluginManager` which is used to define how the :class:`~.PassManager`
+for the stage will be constructed. For example, to create a ``layout`` stage plugin that just
+runs :class:`~.VF2Layout`::
+
+    from qiskit.transpiler.preset_passmanagers.plugin import PassManagerStagePlugin
+    from qiskit.transpiler.preset_passmanagers import common
+    from qiskit.transpiler import PassManager
+    from qiskit.transpiler.passes import VF2Layout
+
+    class VF2LayoutPlugin(PassManagerStagePlugin):
+
+        def pass_manager(self, pass_manager_config):
+            layout_pm = PassManager([VF2Layout(
+                coupling_map=pass_manager_config.coupling_map,
+                properties=pass_manager_config.backend_properties,
+                target=pass_manager_config.target
+            )])
+            layout_pm += common.generate_embed_passmanager(pass_manager_config.coupling_map)
+            return layout_pm
+
+The second step is to expose the :class:`~.PassManagerStagePluginManager`
+subclass as a setuptools entry point in the package metadata. This can be done
+by simply adding an ``entry_points`` entry to the ``setuptools.setup`` call in
+the ``setup.py`` or the plugin package with the necessary entry points under the
+appropriate namespace for the stage your plugin is for. You can see the list
+of stages, entrypoints, and expectations from the stage in :ref:`stage_table`.
+For example, continuing from the example plugin above::
+
+    entry_points = {
+        'qiskit.transpiler.layout': [
+            'vf2 = qiskit_plugin_pkg.module.plugin:VF2LayoutPlugin',
+        ]
+    },
+
+(note that the entry point ``name = path`` is a single string not a Python
+expression). There isn't a limit to the number of plugins a single package can
+include as long as each plugin has a unique name. So a single package can
+expose multiple plugins if necessary. Refer to :ref:`stage_table` for a list
+of reserved names for each stage.
+
```

Comment:
I have tried to follow this very example of creating plugins, adding `qiskit.transpiler.layout` to `entry_points`; now `list_stage_plugins('layout')` includes `vf2`. However, running `transpile(qc, layout_method="vf2")` leads to an error, as level1.py has the code
```
    if layout_method == "trivial":
        _improve_layout = TrivialLayout(coupling_map)
    elif layout_method == "dense":
        _improve_layout = DenseLayout(coupling_map, backend_properties, target=target)
    elif layout_method == "noise_adaptive":
        _improve_layout = NoiseAdaptiveLayout(backend_properties)
    elif layout_method == "sabre":
        _improve_layout = SabreLayout(coupling_map, max_iterations=2, seed=seed_transpiler)
    else:
        raise TranspilerError("Invalid layout method %s." % layout_method)
```
which ignores non-default layouts. Is this expected for this PR, or am I not following the instructions correctly?

Reply:
Good catch it's a bug, I missed that exception in the code. It's just on layout stage plug-ins, we need to drop the else raise from the level1 source (along with the other levels too probably).

Reply:
This should be fixed now with: https://github.com/Qiskit/qiskit-terra/pull/8305/commits/62b4f3f50718e0d28c9d8d5c23c1e9aaff5f9783